collections/src/commonMain/kotlin/AtomicMap.kt

```diff
@@ -0,0 +1,94 @@
+package com.juul.tuulbox.collections
+
+import kotlinx.collections.immutable.ImmutableCollection
+import kotlinx.collections.immutable.ImmutableMap
+import kotlinx.collections.immutable.ImmutableSet
+import kotlinx.collections.immutable.PersistentMap
+import kotlinx.collections.immutable.mutate
+import kotlinx.collections.immutable.persistentHashMapOf
+import kotlinx.collections.immutable.persistentMapOf
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.getAndUpdate
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.flow.updateAndGet
+
+/** Returns an [AtomicMap] that guarantees preservation of iteration order, but may be slower. */
+public fun <K, V> atomicMapOf(
+    vararg pairs: Pair<K, V>,
+): AtomicMap<K, V> = AtomicMap(persistentMapOf(*pairs))
+
+/** Returns an [AtomicMap] that does not guarantee preservation of iteration order, but may be faster. */
+public fun <K, V> atomicHashMapOf(
+    vararg pairs: Pair<K, V>,
+): AtomicMap<K, V> = AtomicMap(persistentHashMapOf(*pairs))
+
+/**
+ * A [Map] that allows for thread safe, atomic mutation. Returned collections such as [entries] and
+ * [iterator] reference a [snapshot] of when they were accessed, and are not mutated when the map is.
+ *
+ * Although mutable, this class intentionally does not implement [MutableMap]. Mutation must use
+ * designated mutator functions ([mutate], [snapshotAndMutate], [mutateAndSnapshot]).
+ */
+public class AtomicMap<K, V> private constructor(
+    @PublishedApi
+    internal val state: MutableStateFlow<PersistentMap<K, V>>,
+) : Map<K, V> {
+
+    internal constructor(initial: PersistentMap<K, V>) : this(MutableStateFlow(initial))
+
+    /** Returns this map as a [StateFlow]. Each mutation will cause a new emit on this flow. */
```

Comment:
English is hard, not sure if "emission" is more correct?

```suggestion
    /** Returns this map as a [StateFlow]. Each mutation will cause a new emission on this flow. */
```